### ISSUE TITLE
Silent 2x inference slowdown when video height exceeds max_height in config; training_config.json is ignored

### ISSUE THREAD

**Issue by yitongxia-1180** [talmolab/sleap#2682]
<html>
<body>
<h3 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Summary</h3>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">I observed a ~2x inference speed difference (38 fps vs 20 fps) on visually identical videos recorded by the same device with the same software at the same time. After extensive investigation, I traced it to a silent performance fallback related to the <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> / <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_width</code> settings in the model config files. There are also two config files in the model folder, and only one of them appears to actually affect inference behavior, which is confusing.</p>
<h3 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Environment</h3>
<ul class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-disc flex flex-col gap-1 pl-8 mb-3">
<li class="whitespace-normal break-words pl-2">sleap-nn 0.1.1</li>
<li class="whitespace-normal break-words pl-2">PyTorch 2.10.0+cu130, CUDA 13.0</li>
<li class="whitespace-normal break-words pl-2">Windows, single GPU</li>
<li class="whitespace-normal break-words pl-2">Bottom-up model</li>
</ul>
<h3 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Observations</h3>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">The model folder contains two config files:</p>
<ul class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-disc flex flex-col gap-1 pl-8 mb-3">
<li class="whitespace-normal break-words pl-2"><code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">initial_config.json</code></li>
<li class="whitespace-normal break-words pl-2"><code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">training_config.json</code></li>
</ul>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">Both contain <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> and <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_width</code> parameters. Both were set to <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height: 1656</code>, <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_width: 1696</code> after training (auto-derived from the training set, presumably).</p>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">I have multiple videos recorded under identical conditions, but with slightly different resolutions (the recording software seems to produce small ROI variations across channels). Inference fps depends entirely on whether the video height fits within <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code>:</p>
<div class="overflow-x-auto w-full px-2 mb-6">
Video resolution (W × H) | Inference speed
-- | --
1664 × 1656 | 38 fps
1696 × 1656 | 38 fps
1672 × 1634 | 20 fps
1648 × 1644 | 20 fps
1680 × 1686 | 20 fps

</div>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">GPU utilization is ~80% on the fast videos and ~45% on the slow ones, suggesting the slow path is starving the GPU.</p>
<h3 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">What I tried</h3>
<ol class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-decimal flex flex-col gap-1 pl-8 mb-3">
<li class="whitespace-normal break-words pl-2"><strong>Padded a slow video to a 32-aligned shape (1728 × 1664)</strong> — still 20 fps. So this is not a cuDNN alignment issue.</li>
<li class="whitespace-normal break-words pl-2"><strong>Cropped/padded a slow video to exactly match a fast video's resolution (1664 × 1656)</strong> — became 38 fps.</li>
<li class="whitespace-normal break-words pl-2"><strong>Edited <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">training_config.json</code> to raise <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> to 1686</strong> (to fit a 1680 × 1686 video) — no effect, still 20 fps.</li>
<li class="whitespace-normal break-words pl-2"><strong>Edited <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">initial_config.json</code> to raise <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> to 1686</strong> — became 38 fps.</li>
<li class="whitespace-normal break-words pl-2"><strong>Edited <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">initial_config.json</code> to set <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height = 2000</code></strong> — still 38 fps. So raising the value arbitrarily seems safe.</li>
</ol>
<h3 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Issues this raises</h3>
<ol class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-decimal flex flex-col gap-1 pl-8 mb-3">
<li class="whitespace-normal break-words pl-2"><strong>Silent performance fallback.</strong> When a video's height exceeds <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code>, inference silently runs ~2x slower with no warning or log message. This is very hard to diagnose — for a long time I thought it was a hardware, encoding, or content issue. At minimum, sleap-nn should emit a warning like:
<blockquote class="ml-2 border-l-4 border-border-300/10 pl-4 text-text-300">
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]"><code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">Input height (1686) exceeds max_height (1656); falling back to non-optimized inference path.</code></p>
</blockquote>
</li>
<li class="whitespace-normal break-words pl-2"><strong><code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">training_config.json</code> appears to be ignored at inference time.</strong> Only <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">initial_config.json</code> actually controls inference behavior. Having two files with the same parameters where only one matters is confusing and error-prone. Either the unused file should be removed, or both should be respected, or it should be clearly documented which one inference reads.</li>
<li class="whitespace-normal break-words pl-2"><strong>Auto-derived <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> / <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_width</code> are too tight.</strong> They appear to be set to the exact max dimensions seen in the training set, which means any new video slightly larger triggers the slow path. A safer default would be to round up generously, or at least document that users should manually raise these values for inference on out-of-distribution-sized videos.</li>
</ol>
<h3 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Suggested fixes (in order of effort)</h3>
<ul class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-disc flex flex-col gap-1 pl-8 mb-3">
<li class="whitespace-normal break-words pl-2">Add a warning when input dimensions exceed <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> / <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_width</code>.</li>
<li class="whitespace-normal break-words pl-2">Clarify in docs which config file is read at inference time, and remove or unify the redundant one.</li>
<li class="whitespace-normal break-words pl-2">Consider making the fast inference path tolerant of larger inputs, or auto-adjust <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_height</code> / <code class="bg-text-200/5 border border-0.5 border-border-300 text-danger-000 whitespace-pre-wrap rounded-[0.4rem] px-1 py-px text-[0.9rem]">max_width</code> at inference based on the actual videos.</li>
</ul>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">Happy to provide more details if useful.</p>
</body>
</html>

## System diagnostics

<details><summary>Output of <code>sleap doctor</code></summary>

```
SLEAP System Diagnostics
========================
Generated: 2026-01-30 12:00:00

[Platform]
  OS:        Darwin 24.6.0
  Platform:  macOS-15.7.3-arm64-arm-64bit-Mach-O
  ...

[Python]
  Version:     3.13.7
  ...

[GPU / CUDA]
  PyTorch:       v2.10.0 (MPS)
  ...

[Packages]
  sleap:          v1.6.0
  sleap-io:       v0.6.3
  sleap-nn:       v0.1.0a4
  ...
```

</details>

<details><summary>Logs</summary>

```
# paste relevant logs here, if any
```

</details>

**Comment by yitongxia-1180** [comment]
Note: I have yet not checked the output of inference at different max_height values, e.g. if super big values would cause the decrease of accuracy or crashing. But anyway, inferring at a low speed would not significantly reduce the accuracy.

**Comment by gitttt-1234** [comment]
Hi @yitongxia-1180,

At inference, we resize each frame to the `max_height` × `max_width` saved in the training config. We do this on purpose: the model was trained at that resolution, so feeding it the same input shape generally gives better accuracy.

The catch is that this resize happens on the CPU (in `apply_sizematcher`, called per-frame in the data pipeline), and when a real interpolation is required, that CPU step starves the GPU, which explains the GPU utilization drop.

The reason your "fast" 1664 × 1656 video is fast is subtle: when the height already matches `max_height`, the aspect-preserving scale factor lands at `eff_scale = 1.0`, so the call collapses to **padding only** and thus no interpolation. That's the same code path, just with the resize op skipped. Any other shape (height larger or smaller than `max_height`) triggers a real bilinear resize on every frame, and that's what costs you the 2x.

**Workaround — use the CLI override**

You don't need to edit any config files. The CLI accepts `--max_height` and `--max_width` flags that override the values from `training_config.yaml`:

```bash
sleap track --model_paths PATH/TO/MODEL --max_height 2000 --max_width 2000 -o out.slp INPUT.mp4
```

or using sleap-nn
```bash
sleap-nn track --model_paths PATH/TO/MODEL --max_height 2000 --max_width 2000 -o out.slp INPUT.mp4
```

The trade-off is that at very different input sizes the model may lose some accuracy since it was trained at a fixed resolution. For small differences like yours (a few percent), this should work.

> training_config.json appears to be ignored at inference time. Only initial_config.json actually controls inference behavior. Having two files with the same parameters where only one matters is confusing and error-prone. Either the unused file should be removed, or both should be respected, or it should be clearly documented which one inference reads.

Reg the two config files, both files in the checkpoint folder are saved on purpose, as part of capturing the full training workflow for reproducibility:

- `initial_config.yaml` — the config you submitted to training (preserved as-is)
- `training_config.yaml` — the resolved/auto-derived config that was actually used (this is the one inference reads)

For legacy SLEAP ≤1.4 models, this is `training_config.json` instead. `initial_config` is never read by any inference path, so edits to it shouldn't change inference behavior. If you want to control inference, the CLI flags above are the right way.

We'll add a warning note in our pipeline whenever the resize happens. Let us know if you have any questions!

Thanks,

Divya
